### PR TITLE
Add programatic defaults

### PR DESCRIPTION
**Pull request by RichardLitt**
Also fixed up a few spelling errors and mistypes, here and there. Got everything I could find.

License: MIT
Signed-off-by: Richard Littauer <richard.littauer@gmail.com>

### CODE REVIEW

File: core/commands/active.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ Lists running and recently run commands.
 		res.SetOutput(req.InvocContext().ReqLog.Report())
 	},
 	Options: []cmds.Option{
-		cmds.BoolOption("v", "verbose", "print more verbose output"),
+		cmds.BoolOption("v", "verbose", "print more verbose output").Default(false),
```
Comment (whyrusleeping) [inline core/commands/active.go]: Is it necessary to add `Default(false)` to every boolean option? I think it can be inferred that the default value is false.

File: core/commands/external.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ func ExternalBinary() *cmds.Command {
 
 			err = cmd.Start()
 			if err != nil {
-				res.SetError(fmt.Errorf("failed to start subcommand: %s", err), cmds.ErrNormal)
+				res.SetError(fmt.Errorf("Failed to start subcommand: %s", err), cmds.ErrNormal)
```
Comment (whyrusleeping) [inline core/commands/external.go]: the convention for errors in go is to use lower case and not capitalize things.

File: core/commands/files/files.go [hunk 3]
```diff
@@ -249,10 +249,10 @@ Examples:
 `,
 	},
 	Arguments: []cmds.Argument{
-		cmds.StringArg("path", false, false, "Path to show listing for. Defaults to '/'."),
+		cmds.StringArg("path", false, false, "Path to show listing for.").Default('/'),
```
Comment (whyrusleeping) [inline core/commands/files/files.go]: strings in go use double quotes. The single quote here will be interpreted as a character and fail at runtime if someone uses the default values.

Comment (whyrusleeping) [inline core/commands/files/files.go]: Also, i just realize this is an argument. Arguments don't have the fancy `Default` decorator yet

File: core/commands/files/files.go [hunk 4]
```diff
@@ -668,7 +668,7 @@ are run with the '--flush=false'.
 `,
 	},
 	Arguments: []cmds.Argument{
-		cmds.StringArg("path", false, false, "Path to flush. Default: '/'."),
+		cmds.StringArg("path", false, false, "Path to flush.").Default("/"),
```
Comment (whyrusleeping) [inline core/commands/files/files.go]: no `.Default` on arguments

File: core/commands/get.go [hunk 5]
```diff
@@ -40,10 +40,10 @@ may also specify the level of compression by specifying '-l=<1-9>'.
 		cmds.StringArg("ipfs-path", true, false, "The path to the IPFS object(s) to be outputted.").EnableStdin(),
 	},
 	Options: []cmds.Option{
-		cmds.StringOption("output", "o", "The path where the output should be stored."),
-		cmds.BoolOption("archive", "a", "Output a TAR archive. Default: false."),
-		cmds.BoolOption("compress", "C", "Compress the output with GZIP compression. Default: false."),
-		cmds.IntOption("compression-level", "l", "The level of compression (1-9). Default: -1."),
+		cmds.StringOption("output", "o", "The path where the output should be stored.").Default("'./<ipfs-path>'"),
```
Comment (whyrusleeping) [inline core/commands/get.go]: lets not do this one. `'./<ipfs-path>'` is not a valid value (if it gets passed it, things will break).

File: core/commands/get.go [hunk 6]
```diff
@@ -40,10 +40,10 @@ may also specify the level of compression by specifying '-l=<1-9>'.
...
+		cmds.BoolOption("archive", "a", "Output a TAR archive.").Default(false),
+		cmds.BoolOption("compress", "C", "Compress the output with GZIP compression.").Default(false),
+		cmds.IntOption("compression-level", "l", "The level of compression (1-9).").Default("-1"),
```
Comment (whyrusleeping) [inline core/commands/get.go]: since this one is an `IntOption` the value you pass to `Default` also needs to be an int, drop the quotes here

File: core/commands/ipns.go [hunk 7]
```diff
@@ -41,11 +41,12 @@ Resolve the value of another name:
 	},
 
 	Arguments: []cmds.Argument{
-		cmds.StringArg("name", false, false, "The IPNS name to resolve. Defaults to your node's peerID.").EnableStdin(),
+		cmds.StringArg("name", false, false, "The IPNS name to resolve.").EnableStdin().Default("your node's peerID"),
```
Comment (whyrusleeping) [inline core/commands/ipns.go]: no default on arguments

File: core/commands/object/object.go [hunk 8]
```diff
@@ -403,7 +403,7 @@ Available templates:
 `,
 	},
 	Arguments: []cmds.Argument{
-		cmds.StringArg("template", false, false, "Template to use. Optional."),
+		cmds.StringArg("template", false, false, "Template to use.").Default("false"),
```
Comment (whyrusleeping) [inline core/commands/object/object.go]: no default on arguments. This argument doesnt even appear to have a default value in the first place, its optional

File: core/commands/pin.go [hunk 9]
```diff
@@ -15,6 +16,14 @@ import (
 	context "gx/ipfs/QmZy2y8t9zQH2a1b8q2ZSLKp17ATuJoCNxxyMFG5qFExpt/go-net/context"
 )
 
+var (
+	tDirect    = "direct"
+	tIndirect  = "indirect"
+	tRecursive = "recursive"
+	tAll 			 = "all"
```
Comment (whyrusleeping) [inline core/commands/pin.go]: run go fmt on all of these, the formatting is off at least here (maybe elsewhere too)

File: core/commands/ping.go [hunk 10]
```diff
@@ -41,7 +41,7 @@ trip latency information.
 		cmds.StringArg("peer ID", true, true, "ID of peer to be pinged.").EnableStdin(),
 	},
 	Options: []cmds.Option{
-		cmds.IntOption("count", "n", "Number of ping messages to send."),
+		cmds.IntOption("count", "n", "Number of ping messages to send.").Default(10),
```
Comment (whyrusleeping) [inline core/commands/ping.go]: since we're setting the default value here, we should drop the logic in the `Run` function for manually setting the default value. The `if found {...` bits

File: core/commands/publish.go [hunk 11]
```diff
@@ -50,13 +50,13 @@ Publish an <ipfs-path> to another public key (not implemented):
...
-		cmds.BoolOption("resolve", "Resolve given path before publishing (default=true)."),
-		cmds.StringOption("lifetime", "t", `Time duration that the record will be valid for. Default: 24h.
+		cmds.BoolOption("resolve", "Resolve given path before publishing.").Default(true),
+		cmds.StringOption("lifetime", "t", `Time duration that the record will be valid for.
 
     This accepts durations such as "300s", "1.5h" or "2h45m". Valid time units are
     "ns", "us" (or "µs"), "ms", "s", "m", "h".
-		`),
-		cmds.StringOption("ttl", "Time duration this record should be cached for (caution: experimental)."),
+		`).Default("24h"),
```
Comment (whyrusleeping) [inline core/commands/publish.go]: again here, if we're setting the defaults at this level, we should clean up the old logic in the `Run` function

File: core/commands/refs.go [hunk 12]
```diff
@@ -49,10 +49,10 @@ Note: List all references recursively by using the flag '-r'.
 		cmds.StringArg("ipfs-path", true, true, "Path to the object(s) to list refs from.").EnableStdin(),
 	},
 	Options: []cmds.Option{
-		cmds.StringOption("format", "Emit edges with given format. Available tokens: <src> <dst> <linkname>."),
-		cmds.BoolOption("edges", "e", "Emit edge format: `<from> -> <to>`."),
-		cmds.BoolOption("unique", "u", "Omit duplicate refs from output."),
-		cmds.BoolOption("recursive", "r", "Recursively list links of child nodes."),
+		cmds.StringOption("format", "Emit edges with given format. Available tokens: <src> <dst> <linkname>.").Default("<dst>"),
```
Comment (whyrusleeping) [inline core/commands/refs.go]: cleanup logic

File: core/commands/stat.go [hunk 13]
```diff
@@ -33,10 +33,10 @@ var statBwCmd = &cmds.Command{
 		ShortDescription: ``,
 	},
 	Options: []cmds.Option{
-		cmds.StringOption("peer", "p", "Specify a peer to print bandwidth for."),
-		cmds.StringOption("proto", "t", "Specify a protocol to print bandwidth for."),
-		cmds.BoolOption("poll", "Print bandwidth at an interval. Default: false."),
-		cmds.StringOption("interval", "i", "Time interval to wait between updating output."),
+		cmds.StringOption("peer", "p", "Specify a peer to print bandwidth for.").Default("overall bandwidth"),
+		cmds.StringOption("proto", "t", "Specify a protocol to print bandwidth for.").Default("all protocols"),
```
Comment (whyrusleeping) [inline core/commands/stat.go]: neither of these values are 'valid options'. If i run `ipfs stats bw --proto="all protocols"` its not going to work.

Setting the default option value here is going to require a bit more in-depth logic changes in the code. We will probably need to allow the user to specify `"all"`

File: core/commands/stat.go [hunk 14]
```diff
@@ -33,10 +33,10 @@ var statBwCmd = &cmds.Command{
...
+		cmds.BoolOption("poll", "Print bandwidth at an interval.").Default(false),
+		cmds.StringOption("interval", "i", "Time interval to wait between updating output.").Default("1s"),
```
Comment (whyrusleeping) [inline core/commands/stat.go]: cleanup `found` logic

File: core/commands/tour.go [hunk 15]
```diff
@@ -25,7 +25,7 @@ IPFS very quickly. To start, run:
 	},
 
 	Arguments: []cmds.Argument{
-		cmds.StringArg("id", false, false, "The id of the topic you would like to tour."),
+		cmds.StringArg("id", false, false, "The id of the topic you would like to tour.").Default("The last accessed tour"),
```
Comment (whyrusleeping) [inline core/commands/tour.go]: no defaults on arguments

File: core/commands/active.go (hunk 1 above)
Comment (RichardLitt) [inline core/commands/active.go]: They're not always false. We could specify only when default is true? cc @jbenet

File: core/commands/external.go (hunk 2 above)
Comment (RichardLitt) [inline core/commands/external.go]: Sounds good.

File: core/commands/files/files.go (hunk 3 above)
Comment (RichardLitt) [inline core/commands/files/files.go]: Can we add it for optional arguments? I believe those should have them.

File: core/commands/get.go (hunk 5 above)
Comment (RichardLitt) [inline core/commands/get.go]: However, it does default to the config-set IPFS path. How should we specify that? That should be in the documentation, somewhere.

File: core/commands/object/object.go (hunk 8 above)
Comment (RichardLitt) [inline core/commands/object/object.go]: Can we say have a note saying "Default is empty"?

File: core/commands/pin.go (hunk 9 above)
Comment (RichardLitt) [inline core/commands/pin.go]: Is there a quick way to run `go fmt` on all changed files from a commit?

File: core/commands/ping.go (hunk 10 above)
Comment (RichardLitt) [inline core/commands/ping.go]: Oh! Default actually uses the value passed, and doesn't just format it as text? Ok, that makes things more complex. I'll need to check this over, then.

File: core/commands/pin.go (hunk 9 above)
Comment (RichardLitt) [inline core/commands/pin.go]: Nvm. Got it! Thanks tmg

File: core/commands/files/files.go (hunk 3 above)
Comment (whyrusleeping) [inline core/commands/files/files.go]: we can add it, but that will be in another PR

**Comment by whyrusleeping** [comment]
@RichardLitt ping.

File: core/commands/active.go (hunk 1 above)
Comment (RichardLitt) [inline core/commands/active.go]: Going to assume that yes, we should add `Default(false)` from now on. See https://github.com/ipfs/go-ipfs/pull/2582#issuecomment-214925152

File: core/commands/refs.go (hunk 12 above)
Comment (RichardLitt) [inline core/commands/refs.go]: Not sure how - this logic looks right to me. Is this example wrong? 

```
		edges, _, err := req.Option("edges").Bool()
		if err != nil {
			res.SetError(err, cmds.ErrNormal)
			return
		}
```

File: core/commands/stat.go (hunk 13 above)
Comment (RichardLitt) [inline core/commands/stat.go]: Not sure I know enough to do that. Removing it, for now.

**Comment by RichardLitt** [comment]
Did what I could. Take another pass?

**Comment by whyrusleeping** [comment]
@RichardLitt I'm not comfortable merging this. It doesnt compile, and there are tons of little mistakes, such as enabling unrestricted api access by default, setting 'Defaults' with the wrong type (a bool on a string option, etc).

If you want to get these in, do them in smaller, well thought out patches. I don't want to accidentally introduce incorrect behaviour.
